python_arg_parser to allow fake tensor element in symint_list when in dynamo mode  #95424

lantiankaikai opened the pull request:
Failing mechanism on #95424 :
In dynamo mode, when passing numpy.int_ to 'shape' like param (Sequence[Union[int, symint]]) is wrapped as list with FakeTensor.  However, in python_arg_parser, parser expect int in symint_list but got FakeTensor.

Following #85759, this PR allow tensor element in symint_list when in dynamo mode

This PR also fix below test with similar failing mechanism 
pytest ./generated/test_huggingface_diffusers.py -k test_016 
pytest ./generated/test_ustcml_RecStudio.py -k test_036 

Fixes #ISSUE_NUMBER


cc @ezyang @gchanan @zou3519 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @anijain2305 @soumith @desertfire

lantiankaikai committed:
python_arg_parser to allow tensor when varargs_eligible is true  #95424

Failing mechanism on #95424 :
In dynamo mode, when passing numpy.int_ to 'shape' like param (Sequence[Union[int, symint]]) is wrapped as FakeTensor.  However, in python_arg_parser, parser expect a symint_list but input is a FakeTensor.

Following #85759, this PR allow tensor as input when varargs_eligible is true (function with single int varargs param)

lantiankaikai committed:
fix lint #95424

ezyang commented:
It's a good sign that CI hasn't failed yet though.

lantiankaikai committed:
Merge branch 'pytorch:viable/strict' into viable/strict

ezyang commented:
@eellison can you please review this

eellison commented:
Yes will review

lantiankaikai committed:
add test that allowing fake tensor to be pass into symint_list type in dynamo

eellison commented on test/dynamo/test_backends.py:
```diff
@@ -212,6 +213,25 @@ def test_default_mps_to_aot_eager(self):
+        x = torch.rand([4, 4])
+        model = MyModule()
+        orig_out = model(x)
+        opt_model = torch._dynamo.optimize("eager")(MyModule())
```
What happens when you add `nopython=True` to this call ?

lantiankaikai replied:
It actually failed with nopython=True
```
RuntimeError: Failed running call_method reshape(*(FakeTensor(FakeTensor(..., device='meta', size=(4, 4)), cpu), FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu)), **{}):
aten._local_scalar_dense.default
```

lantiankaikai replied:
Did more experiment by giving reshape with combination of int and np.int

```
class MyModule(torch.nn.Linear):
        def __init__(self):

            super().__init__(np.int_(4), np.int_(4))   
            # super().__init__(4, np.int_(4))   

        def forward(self, x):
            return x.reshape(self.in_features, self.out_features)  # (np.int, np.int) or (int, np.int)

x = torch.rand([4, 4])
model = MyModule()
opt_model = torch._dynamo.optimize("eager", nopython=True)(MyModule())
print(opt_model(x))

```
case 1: original code w/ nopython=False
(np.int, np.int) ==>  
`reshape(): argument 'shape' (position 1) must be tuple of ints, but found element of type FakeTensor at pos 0`
(int, np.int)          ==> **okay** , since varargs_eligible and is_int_or_symint_list is checking only 1st element.

case 2: original code w/ nopython=True
(np.int, np.int) ==> `reshape(): argument 'shape' (position 1) must be tuple of ints, but found element of type FakeTensor at pos 0`
(int, np.int)          ==> 
`
RuntimeError: Failed running call_method reshape(*(FakeTensor(FakeTensor(..., device='meta', size=(4, 4)), cpu), 4, FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu)), **{}):
                           aten._local_scalar_dense.default
`

case 3: this PR w/ nopython=False
(np.int, np.int) ==> **okay**
(int, np.int)          ==> **okay**

case 4: this PR w/ nopython=True
(np.int, np.int) ==> 
```
RuntimeError: Failed running call_method reshape(*(FakeTensor(FakeTensor(..., device='meta', size=(4, 4)), cpu), FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu), FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu)), **{}):
aten._local_scalar_dense.default
```
(int, np.int)  ==> 
```
RuntimeError: Failed running call_method reshape(*(FakeTensor(FakeTensor(..., device='meta', size=(4, 4)), cpu), 4, FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu)), **{}):
aten._local_scalar_dense.default
```

eellison replied:
Graph break is fine here - just wanted to see what the behavior was

eellison reviewed (commented):
What path in the arg parser gets hit when you do :

```
x = torch.rand([4, 4])
y = torch.tensor(16)
x.reshape(y)
```

?

lantiankaikai committed:
add is_dynamo_compiling in eval_frame.c and update python_arg_parser

lantiankaikai committed:
Merge branch 'pytorch:viable/strict' into viable/strict

lantiankaikai committed:
Merge branch 'viable/strict' of https://github.com/lantiankaikai/pytorch into viable/strict

lantiankaikai committed:
fix lint

lantiankaikai commented:
@eellison 
```
x = torch.rand([4, 4])
y = torch.tensor(16)
x.reshape(y)
```
same path in arg_parser (is_int_or_symint_list), but this case is okay since y is a tensor so it get implicit convert to int. But if y = np.int_(16), it will failed, I'm guessing since python think np.int and int are the same type so no implicit conversion there. 

however, interestingly if pass y as param, same issue can be repro 
```
def f(y)
  x = torch.rand([4, 4])
  x.reshape(y)

opt_f = torch._dynamo.optimize("eager")(f)
opt_f(torch.tensor(16))
```

Further discussed w/ @yanboliang and suggest to add a flag to check if in dynamo mode before allow fake tensor. WIP

lantiankaikai committed:
fix lint

lantiankaikai committed:
in dynamo allow fake tensor to act as single int param when param type is symint_list

lantiankaikai committed:
move the check to is_int_or_symint

lantiankaikai committed:
lint & skip size1 check when in dynamo mode

lantiankaikai committed:
add more test for param type int_or_vararg_int

lantiankaikai committed:
revert allow fake tensor to act as single int param when param type is int_or_symint_list

it broke the test when function's input overload with Tensor, Union[_int, SymInt], or Sequence[Union[_int, SymInt]], like tensor_split

Separate the issue to another PR

ezyang commented:
@yanboliang @eellison any update on this?

yanboliang commented:
@ezyang The conclusion so far is: when ```is_dynamo_compiling```, we allow scalar tensors to act as ints in an intlist argument, which is exactly the same as we did for ```jit::tracer::isTracing```. This fixed several similar arg mismatch issues in 14k github models and passed all CI checks(except the building link errors). I think we can get this in after @lantiankaikai addressed the deploy/xla building failures.

lantiankaikai committed:
add dynamo_core_headers

lantiankaikai commented:
@pytorchbot label "topic: not user facing"

lantiankaikai committed:
Merge branch 'pytorch:viable/strict' into viable/strict

yanboliang commented:
@lantiankaikai Can you re-enable the following unit test? It seems this PR fix that issue.
https://github.com/pytorch/pytorch/blob/e9a7115605ee5b6ae38ea5716a4abea5aa415333/test/dynamo/test_unspec.py#L111-L113

yanboliang commented:
This PR added ```is_dynamo_compiling``` in ```torch/csrc/dynamo/eval_frame.h```, but XLA and torch deploy tests are failing due to:
```
2023-05-19T07:26:01.7529883Z /opt/conda/lib/python3.8/site-packages/torch/include/torch/csrc/utils/python_arg_parser.h:57:10: fatal error: 'torch/csrc/dynamo/eval_frame.h' file not found
2023-05-19T07:26:01.7530496Z #include <torch/csrc/dynamo/eval_frame.h>
2023-05-19T07:26:01.7530961Z          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2023-05-19T07:26:01.7830415Z 1 error generated.
2023-05-19T07:26:01.8190228Z CMakeFiles/test_ptxla.dir/build.make:243: recipe for target 'CMakeFiles/test_ptxla.dir/test_xla_backend_intf.cpp.o' failed
```
Can you suggest how to fix it? @malfet

lantiankaikai committed:
Update test_unspec.py

malfet committed:
Install eval_frame.h

malfet commented:
> Can you suggest how to fix it?

@yanboliang both deploy and XLA take installed copy of pytorch, i.e. one needs to make sure that `eval_frame.h` is part of PyTorch wheel/conda package, by mentioning it in setup.py, what I've attempted to do in https://github.com/lantiankaikai/pytorch/commit/81614f2bc498cca563c016775eba82d2369b6d44

lantiankaikai committed:
Revert "add dynamo_core_headers"

This reverts commit 9a5b2e3808ace152a1309221bccbc1026ede570d.

lantiankaikai committed:
Merge branch 'pytorch:viable/strict' into viable/strict

yanboliang commented on test/dynamo/test_backends.py:
```diff
@@ -215,6 +217,27 @@ def test_mps_not_supported(self):
         )
 
 
+class TestDynamoModeArgParser(torch._dynamo.test_case.TestCase):
```
This is not a backend test, let's move it to ```test/dynamo/test_misc.py```

yanboliang commented:
@ezyang @eellison This looks good to me, can you help to double check? Thanks.

lantiankaikai committed:
move unit test to test_misc.py

lantiankaikai committed:
Merge branch 'viable/strict' of https://github.com/lantiankaikai/pytorch into viable/strict

lantiankaikai committed:
lint

lantiankaikai committed:
lint
